File: src/main/java/org/gangel/graphomance/engine/TestLimit.java
package org.gangel.graphomance.engine;

import java.time.Duration;

public class TestLimit {
    private final long maxIterations;
    private final Duration minDuration;
    private final Duration maxDuration;
    private long startTime;
    private long cnt;
    private long warmUpIterations = 10;

    public TestLimit(long maxIterations, Duration minDuration, Duration maxDuration) {
        this.minDuration = minDuration;
        this.maxIterations = maxIterations;
        this.maxDuration = maxDuration;
        startTime = System.currentTimeMillis();
    }

    public void increment() {
        ++cnt;
    }

    public boolean isWarmUpPhase() {
        return (cnt < warmUpIterations);
    }

    public boolean isDone() {
        if (System.currentTimeMillis() - startTime < minDuration.toMillis()) {
            return false;
        }
        return (cnt + warmUpIterations >= maxIterations) || (System.currentTimeMillis() - startTime >= maxDuration.toMillis());
    }
}

File: src/main/kotlin/org/graphomance/engine/TestLimit.kt
package org.graphomance.engine

import java.time.Duration

class TestLimit(private val maxIterations: Long,
	private val minDuration: Duration,
	private val maxDuration: Duration
) {
	private val startTime: Long = System.currentTimeMillis()
	private var cnt: Long = 0
	private val warmUpIterations: Long = 10
	fun increment() {
		++cnt
	}

	val isWarmUpPhase: Boolean
		get() = cnt < warmUpIterations

	val isDone: Boolean
		get() = if (System.currentTimeMillis() - startTime < minDuration.toMillis()) {
			false
		} else cnt + warmUpIterations >= maxIterations || System.currentTimeMillis() - startTime >= maxDuration.toMillis()

}